src/json.cpp

```diff
@@ -0,0 +1,114 @@
+#include <migraphx/serialize.hpp>
+#include <migraphx/argument.hpp>
+#include <migraphx/literal.hpp>
+#include <nlohmann/json.hpp>
+#include <migraphx/json.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+using json = nlohmann::json;
+
+void value_to_json(const value& val, json& j);
+
+template <class T>
+void value_to_json(const T& x, json& j)
+{
+    j = x;
+}
+
+void value_to_json(const std::vector<value>& x, json& j)
+{
+    for(auto& v : x)
+    {
+        if(v.get_key().empty())
+        {
+            json jj;
+            value_to_json(v, jj);
+            j.push_back(jj);
+        }
+        // in the case of an object
+        else
+        {
+            value_to_json(v, j);
+        }
+    }
+}
+
+template <class T, class U>
+void value_to_json(const std::pair<T, U>& x, json& j)
+{
+    json jj;
+    value_to_json(x.second, jj);
+    j[x.first] = jj;
+}
+
+void value_to_json(std::nullptr_t&, json& j) { j = {}; }
+
+void value_to_json(const value& val, json& j)
+{
+    val.visit([&](auto v) { value_to_json(v, j); });
+}
+
+void to_json(json& j, const value& val) { value_to_json(val, j); }
+
+std::string to_json_string(const value& val)
+{
+    json j;
+    value_to_json(val, j);
+    return j.dump();
+}
+
+migraphx::value value_from_json(const json& j)
+{
+    migraphx::value val;
+    json::value_t type = j.type();
+    switch(type)
+    {
+    case json::value_t::null: val = migraphx::value(); break;
+
+    case json::value_t::boolean: val = j.get<bool>(); break;
+
+    case json::value_t::number_float: val = j.get<double>(); break;
+
+    case json::value_t::number_integer: val = j.get<int64_t>(); break;
+
+    case json::value_t::number_unsigned: val = j.get<uint64_t>(); break;
+
+    case json::value_t::string: val = j.get<std::string>(); break;
+
+    case json::value_t::array:
+        std::transform(j.begin(), j.end(), std::back_inserter(val), [&](const auto& v) {
+            return value_from_json(v);
```

Comment:
I think this can be written as `return v.get<value>()`. Also, I think it would be better to use the variable `x` or `jj`, to make it a littler clearer its `json` object and not `value`.

Reply:
the v.get<value>() has a compilation error as I mentioned before.

Reply:
You need to put the `template` keyword there, or change the lambda from `const auto& v` to `const json& v`.

Reply:
Tried, looks like it does not work. I read more materials, from https://github.com/nlohmann/json#implicit-conversions, It is recommended to NOT USE implicit conversions FROM a JSON value, particularly for complex data types. Maybe the json implementation has some limitations for that?